src/once_cell.rs

```diff
@@ -0,0 +1,58 @@
+use crate::Python;
+use std::cell::UnsafeCell;
+
+/// A write-once cell similar to once_cell::OnceCell.
+///
+/// Unlike once_cell::sync which uses locks to achieve thread safety, this implementation uses
+/// the Python GIL to mediate concurrent access.
+pub struct OnceCell<T>(UnsafeCell<Option<T>>);
+
+// T: Send is needed for Sync because the thread which drops the OnceCell can be different
+// to the thread which fills it.
+unsafe impl<T: Send + Sync> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+impl<T> OnceCell<T> {
+    pub const fn new() -> Self {
+        Self(UnsafeCell::new(None))
+    }
+
+    pub fn get(&self, _py: Python) -> Option<&T> {
+        // Safe because if the cell has not yet been written, None is returned.
+        unsafe { &*self.0.get() }.as_ref()
+    }
+
+    pub fn get_or_init<F>(&self, py: Python, f: F) -> &T
+    where
+        F: FnOnce() -> T,
+    {
+        let inner = unsafe { &*self.0.get() }.as_ref();
+        if let Some(value) = inner {
+            return value;
+        }
+
+        // Note that f() could release the GIL, so it's possible that another thread writes to this
+        // OnceCell before f() finishes. That's fine; we'll just have to discard the value
+        // computed here and accept a bit of wasted computation.
+        let value = f();
```

Comment:
Unfortunately `f()` can call things like `py.allow_threads()` to temporarily release the GIL, or it might call Python APIs which may in turn release the GIL.

Reply:
I'll add the word "temporarily" to the comment. 👍

Reply:
Oh I see, I didn’t know we could do that. Thanks for the explanation!